app/controllers/confirmations_controller.rb

```diff
@@ -1,8 +1,96 @@
 class ConfirmationsController < Devise::ConfirmationsController
+  include AlertContactCreator
+  include NewsletterSubscriber
+
+  before_action :set_minimum_password_length, only: [:show, :confirm]
+  before_action :set_email_types
+
+  helper_method :subscribe_to_alerts?
+
+  def show
+    return head(:ok) if request.head?
+
+    self.resource = resource_class.find_by(confirmation_token: params[:confirmation_token])
+
+    if resource.nil?
+      redirect_to new_user_confirmation_path and return
+    elsif resource.confirmed? && current_user.nil?
+      flash[:error] = I18n.t('errors.messages.already_confirmed')
+      redirect_to new_session_path(resource_name) and return
+    elsif resource.confirmed? && current_user
+      flash[:error] = I18n.t('devise.failure.already_authenticated')
+      redirect_to after_sign_in_path_for(resource) and return
+    else
+      @allow_alerts = allow_alerts?(resource)
+      @can_subscribe_newsletter = can_subscribe_newsletter?(resource)
+      @interests = default_interests(resource) if @can_subscribe_newsletter
+      render :show
+    end
+  end
+
+  def confirm
+    self.resource = resource_class.find_by(confirmation_token: params[:confirmation_token])
+
+    if resource.nil?
+      respond_with_navigational(resource.errors, status: :unprocessable_entity) { render :show }
+      return
+    end
+
+    resource.assign_attributes(
+      confirmed_at: Time.zone.now, # setting confirmed means password validation is now active
+      password: resource_params[:password],
+      password_confirmation: resource_params[:password_confirmation],
+      preferred_locale: resource_params[:preferred_locale] || 'en',
+      terms_accepted: resource_params[:terms_accepted]
+    )
+
+    if resource.valid? && resource.terms_accepted
+      resource.save
+
+      subscribe_to_emails(resource)
+      devise_sign_in(resource)
+
+      flash[:success] = I18n.t('devise.confirmations.confirmed')
+      respond_with resource, location: after_resetting_password_path_for(resource)
+    else
+      @allow_alerts = allow_alerts?(resource)
```

Comment:
These three lines are repeated from the show method above. Can they be standardised into one place?